File: weave_introduction_handson/jp/1_2_agent_sdk.py
"""
1_2: OpenAI Agent SDK - エージェント構築

このスクリプトで学べること:
================================
1. OpenAI Agent SDK の基本的な使い方
2. function_tool を使ったツール定義
3. WeaveTracingProcessor によるトレーシング

実行後に確認する場所:
================================
- Traces タブ: エージェント実行とツール呼び出し
- Code タブ: トレースされた関数と Agent SDK 連携

補足:
================================
このスクリプトでは、OpenAI Agents SDK と Weave の基本的な連携を扱います。
Agent アプリケーションでは、LLM 呼び出しだけでなく、ツール実行や複数ステップの
意思決定をまとめて追跡できることが重要です。

Weave では Agent 向け Trace の新機能が開発されており、現時点では
Public Preview として提供されています。最新の仕様は以下を確認してください。
https://docs.wandb.ai/weave/guides/tracking/trace-agents
"""

import asyncio
from dotenv import load_dotenv
import weave
from agents import Agent, Runner, function_tool, set_trace_processors
from weave.integrations.openai_agents.openai_agents import WeaveTracingProcessor

from config_loader import init_weave

# Load environment variables
load_dotenv()

# Initialize Weave
init_weave()

# WeaveTracingProcessor を設定してトレーシングを有効化
set_trace_processors([WeaveTracingProcessor()])


# =============================================================================
# 1. Agent with Multiple Tools - 複数ツールを持つエージェント
# =============================================================================
print("\n" + "=" * 60)
print("1. Agent with Multiple Tools - 複数ツールを持つエージェント")
print("=" * 60)


@function_tool
def search_database(query: str) -> list:
    """Search the database for relevant information."""
    # モックのデータベース検索
    return [
        {"id": 1, "title": f"Result 1 for: {query}", "score": 0.95},
        {"id": 2, "title": f"Result 2 for: {query}", "score": 0.87},
    ]


@function_tool
def send_email(to: str, subject: str, body: str) -> str:
    """Send an email to a recipient."""
    # モックのメール送信
    return f"Email sent to {to} with subject: {subject}"


research_agent = Agent(
    name="ResearchAssistant",
    instructions="""You are a research assistant. You can:
    - Search the database for information
    - Send emails with findings
    Help the user with their research tasks.""",
    tools=[search_database, send_email],
)


async def run_research(input: str) -> str:
    """Research Agent を実行"""
    response = await Runner.run(research_agent, input)
    return response.final_output


result = asyncio.run(run_research("Search for information about machine learning and email the results to user@example.com"))
print(f"Result: {result[:100]}...")


print("\n" + "=" * 60)
print("Agent SDK Demo Complete!")
print("=" * 60)
print("""
まとめ:
- @function_tool で Agent SDK のツールを定義
- Agent() と Runner.run() でエージェントを実行
- WeaveTracingProcessor でエージェント実行を Weave に記録

Weave UI で確認:
- Traces タブでエージェント実行とツール呼び出しを確認
- Inputs/Outputs で各ツールの入出力を確認
""")

File: weave_introduction_handson_typescript/jp/1_2_agent_sdk.ts
/**
 * 1_2: OpenAI Agents SDK Integration - エージェント構築
 *
 * TypeScript 対応: 対応
 * - weave@0.15.1 以降では OpenAI Agents SDK integration が export されています。
 * - `node --import=weave/instrument ...` で起動すると、@openai/agents が自動計装されます。
 * - このスクリプトでは `instrumentOpenAIAgents()` も呼び、hook が効かない環境でも登録を試みます。
 */

import { Agent, run, tool } from "@openai/agents";
import { z } from "zod";
import * as weave from "weave";

import { getModelName, initWeave, printSection } from "../src/config.js";

await initWeave();

const agentsInstrumented = await weave.instrumentOpenAIAgents();
console.log(`OpenAI Agents instrumentation: ${agentsInstrumented ? "enabled" : "not available"}`);

// =============================================================================
// SECTION 1: Agent SDK Integration Status
// =============================================================================
printSection("1. Agent SDK Integration Status - Agent SDK 連携の状態");

console.log(`
この章では @openai/agents と Weave の TypeScript integration を扱います。

確認ポイント:
- Agent run が Weave Trace として記録される
- tool call が子 span / 子 call として見える
- underlying OpenAI model call も trace される

Python 版と同じく、Agent アプリケーションでは LLM call だけでなく、
tool execution や multi-step workflow の流れを追えることが重要です。
`);

// =============================================================================
// SECTION 2: Define Tools
// =============================================================================
printSection("2. Define Tools - ツール定義");

const searchDatabase = tool({
  name: "search_database",
  description: "Search a mock internal database for relevant research notes.",
  parameters: z.object({
    query: z.string().describe("Search query"),
  }),
  async execute({ query }) {
    return [
      { id: 1, title: `Result 1 for: ${query}`, score: 0.95 },
      { id: 2, title: `Result 2 for: ${query}`, score: 0.87 },
    ];
  },
});

const sendEmail = tool({
  name: "send_email",
  description: "Send a mock email to a recipient.",
  parameters: z.object({
    to: z.string().email().describe("Recipient email address"),
    subject: z.string().describe("Email subject"),
    body: z.string().describe("Email body"),
  }),
  async execute({ to, subject, body }) {
    return {
      status: "sent",
      to,
      subject,
      bodyPreview: body.slice(0, 80),
    };
  },
});

console.log("Defined tools: search_database, send_email");

// =============================================================================
// SECTION 3: Run Agent
// =============================================================================
printSection("3. Run Agent - エージェント実行");

const researchAgent = new Agent({
  name: "ResearchAssistant",
  model: getModelName(),
  instructions: `You are a research assistant.
Use search_database to find relevant information.
If the user asks to email results, call send_email with a concise summary.`,
  tools: [searchDatabase, sendEmail],
});

const result = await run(
  researchAgent,
  "Search for information about machine learning and email the results to user@example.com."
);

console.log(`Final output: ${String(result.finalOutput).slice(0, 300)}...`);

// =============================================================================
// COMPLETE
// =============================================================================
printSection("Agent SDK Integration Demo Complete!");
console.log(`
まとめ:
- @openai/agents の Agent / tool / run を使った workflow を実行
- weave.instrumentOpenAIAgents() で OpenAI Agents integration を有効化
- Weave UI で agent trace、tool call、underlying model call を確認

補足:
- Python では OpenAI Agents SDK、CrewAI、AutoGen、LlamaIndex、LangChain などの integration も利用できます。
- TypeScript で特定の Agent SDK integration が必要な場合は、W&B までご連絡ください。
`);
